### PR TITLE
Block mapping binding in foreach and query for anydata and any type collections

### PR DESCRIPTION
**Pull request by gimantha**
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/29545

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Review by mohanvive — DISMISSED** [review]

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 1]
```diff
@@ -1299,6 +1299,10 @@ void handleDeclaredVarInForeach(BLangVariable variable, BType rhsType, SymbolEnv
                 recursivelySetFinalFlag(tupleVariable);
                 break;
             case RECORD_VARIABLE:
+                if (rhsType.tag == TypeTags.ANYDATA || rhsType.tag == TypeTags.ANY) {
+                    dlog.error(variable.pos, DiagnosticErrorCode.INVALID_RECORD_BINDING_PATTERN, rhsType);
+                    return;
+                }
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:1305]: Should we fail this at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter#validateRecordVariable instead?

By removing the any and anydata cases at https://github.com/ballerina-platform/ballerina-lang/blob/master/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java#L2266?

Comment (gimantha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:1305]: I thought of doing it, but there are multiple usages of  validateRecordVariables. Basically it is used for validating simple assignments to record variables as well

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:1305]: Can you please share an example? Wondering if those should actually result in compilation errors too?

Comment (gimantha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:1305]: @MaryamZi I actually double checked it, We can of course move it to SymbolEnter. One of the negative test was failing with the current changes and that test is  kind of not correct. I will send the changes

File: tests/jballerina-unit-test/src/test/resources/test-src/query/query-negative.bal [hunk 2]
```diff
@@ -420,3 +420,15 @@ public function testInvalidInputType() {
 function testIncompatibleSelectType(stream<string, error> clientStream) returns error? {
     return from string num in clientStream select {a: 1};
 }
+
+function testMapBindingPatternsAnydataType() {
+    map<anydata> keyValsMap = {foo:"sss", bar:"ffff"};
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/query/query-negative.bal:425]:
```suggestion
    map<anydata> keyValsMap = {foo: "sss", bar: "ffff"};
```

File: tests/jballerina-unit-test/src/test/resources/test-src/query/query-negative.bal [hunk 3]
```diff
@@ -420,3 +420,15 @@ public function testInvalidInputType() {
...
+    var x = from var {k} in keyValsMap
+                 select k;
+}
+
+function testMapBindingPatternsAnyType() {
+    map<any> keyValsMap = {foo:"sss", bar:"ffff"};
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/query/query-negative.bal:431]:
```suggestion
    map<any> keyValsMap = {foo: "sss", bar: "ffff"};
```

**Review by MaryamZi — APPROVED** [review]

**Pull request merged**